### PR TITLE
[minor_change] enhanced lag policy support

### LINKED ISSUE #929: aci_l3out_floating_svi does not support enhanced lag option for vmware
The resource does not support the lag policy for a floating SVI.

image

As shown in the image above, the lag policy is created but it is not possible to apply via a aci_l3out_floating_svi  due to no option.

It requires two nested child objects (l3extVirtualLIfPLagPolAtt, l3extRsVSwitchEnhancedLagPol). 

Can be done with `aci_rest` but is not tracked in terms of if the lag policy is deleted or changed, TF does not "see" that change. trying with `aci_rest_managed` does not seem to be easily (or even possible) to achieve.

Example JSON payload from Ansible ACI rest call for a floating svi on l3out with vmware.

```

{
    "l3extVirtualLIfP": {
        "attributes": {
            "addr": "{{l3out_floating_svi_primary_ip}}",
            "annotation": "",
            "autostate": "disabled",
            "descr": "",
            "encap": "vlan-{{allocated_vlan_id}}",
            "encapScope": "local",
            "ifInstT": "ext-svi",
            "ipv6Dad": "enabled",
            "llAddr": "::",
            "mode": "regular",
            "mtu": "1500",
            "name": "",
            "nodeDn": "topology/pod-{{pod_id}}/node-{{node_id}}",
            "targetDscp": "unspecified"
        },
        "children": [
            {
                "l3extRsDynPathAtt": {
                    "attributes": {
                        "annotation": "",
                        "floatingAddr": "{{l3out_floating_ip}}",
                        "forgedTransmit": "Disabled",
                        "macChange": "Disabled",
                        "promMode": "Disabled",
                        "tDn": "uni/vmmp-VMware/dom-{{vm_domain}}"
                    },
                    "children": [
                        {
                            "l3extVirtualLIfPLagPolAtt": {
                                "attributes": {
                                    "annotation": ""
                                },
                                "children": [
                                    {
                                        "l3extRsVSwitchEnhancedLagPol": {
                                            "attributes": {
                                                "annotation": "",
                                                "tDn": "uni/vmmp-VMware/dom-{{vm_domain}}/vswitchpolcont/enlacplagp-{{vm_domain_lag_policy}}"
                                            }
                                        }
                                    }
                                ]
                            }
                        }
                    ]
                }
            },
            {
                "l3extIp": {
                    "attributes": {
                        "addr": "{{l3out_floating_svi_secondary_ip}}",
                        "annotation": "",
                        "descr": "",
                        "ipv6Dad": "enabled",
                        "name": "",
                        "nameAlias": ""
                    }
                }
            }
        ]
    }
}

```

### PR DESCRIPTION
**Pull request by akinross**
fix #929

### CODE REVIEW

File: aci/resource_aci_l3extvirtuallifp.go [hunk 1]
```diff
@@ -246,6 +250,43 @@ func setVirtualLogicalInterfaceProfileAttributes(l3extVirtualLIfP *models.Virtua
...
+
+func getAndSetL3extRsDynPathAttFromLogicalInterfaceProfile(client *client.Client, dn string, d *schema.ResourceData) (*schema.ResourceData, error) {
+	l3extRsDynPathAttData, err := client.ReadRelationl3extRsDynPathAttFromLogicalInterfaceProfile(dn)
+	if err != nil {
+		log.Printf("[DEBUG] Error while reading relation l3extRsDynPathAtt %v", err)
+		d.Set("relation_l3ext_rs_dyn_path_att", make([]interface{}, 0, 1))
+		return nil, err
+	} else {
+		l3extRsDynPathAttMap := l3extRsDynPathAttData.([]map[string]string)
+		st := make([]map[string]string, 0)
```
Comment (sajagana) [inline aci/resource_aci_l3extvirtuallifp.go:273]: You can move the st before the if condition and use l3_ext_rs_dyn_path or some good name instead st.

**Review by sajagana — COMMENTED** [review]
Hi Akini,

1. Data Source attributes do not match with Resource attributes.
2. l3extVirtualLIfP.DistinguishedName to dn in getRemoteVirtualLogicalInterfaceProfile function.
3. Please also take care of the documentation for the data source and resources.

File: aci/resource_aci_l3extvirtuallifp.go (hunk 1 above)
Comment (akinross) [reply aci/resource_aci_l3extvirtuallifp.go:273]: ok

File: examples/l3out_floating_svi/main.tf [hunk 2]
```diff
@@ -56,6 +56,11 @@ resource "aci_l3out_floating_svi" "example" {
     mac_change       = "Enabled"
     promiscuous_mode = "Disabled"
   }
+  relation_l3ext_rs_dyn_path_att {
+    tdn              = aci_vmm_domain.example_vmm.id
+    floating_address = "10.20.30.254/16"
+    enhanced_lag_policy_tdn = "uni/vmmp-VMware/dom-example/vswitchpolcont/enlacplagp-test"
```
[the remark this answers is not in the document]
Comment (akinross) [reply examples/l3out_floating_svi/main.tf:62]: ok fixed indent

**Review by sajagana — DISMISSED** [review]
LGTM!

**Comment by shrsr** [comment]
LGTM
We just need to come to a decision regarding the default value for enhanced lag policy sub attribute.

**Review by xinyuezhao — DISMISSED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]
LGTM

File: examples/l3out_floating_svi/main.tf [hunk 3]
```diff
@@ -30,7 +30,16 @@ resource "aci_logical_node_profile" "node_profile" {
 
 resource "aci_logical_interface_profile" "interface_profile" {
   logical_node_profile_dn = aci_logical_node_profile.node_profile.id
-  name                    = "demo_int_prof"
+  name = "demo_int_prof"
```
Comment (lhercot) [inline examples/l3out_floating_svi/main.tf:33]: Can you run `terraform fmt` on the example files?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/data_source_aci_l3extvirtuallifp.go [hunk 4]
```diff
@@ -97,6 +91,39 @@ func dataSourceAciVirtualLogicalInterfaceProfile() *schema.Resource {
...
+						},
+						"mac_change": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"promiscuous_mode": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"enhanced_lag_policy_tdn": {
```
[the remark this answers is not in the document]
Comment (akinross) [reply aci/data_source_aci_l3extvirtuallifp.go:120]: you are correct, made mistake in naming will correct

File: aci/resource_aci_l3extvirtuallifp.go [hunk 5]
```diff
@@ -193,6 +192,11 @@ func resourceAciVirtualLogicalInterfaceProfile() *schema.Resource {
 							}, false),
 							Default: "Disabled",
 						},
+						"enhanced_lag_policy_tdn": {
```
[the remark this answers is not in the document]
Comment (akinross) [reply aci/resource_aci_l3extvirtuallifp.go:195]: same as datasource will correct

File: examples/l3out_floating_svi/main.tf (hunk 3 above)
Comment (akinross) [reply examples/l3out_floating_svi/main.tf]: ok

File: aci/resource_aci_l3extvirtuallifp.go [hunk 6]
```diff
@@ -246,6 +250,43 @@ func setVirtualLogicalInterfaceProfileAttributes(l3extVirtualLIfP *models.Virtua
 	return d, nil
 }
 
+func getL3extRsVSwitchEnhancedLagPol(client *client.Client, dn, tDn string) string {
+	l3extVirtualLIfPLagPolAttDn := fmt.Sprintf("%s/rsdynPathAtt-[%s]/vlifplagpolatt", dn, tDn)
```
Comment (lhercot) [inline aci/resource_aci_l3extvirtuallifp.go:254]: You use %s/rsdynPathAtt-[%s]/vlifplagpolatt 3 times, can we add it to the go client or somewhere to reuse it?

**Review by lhercot — CHANGES_REQUESTED** [review]
Can you also rebase on latest master branch?

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**